Amazon.IonObjectMapper/IonObjectSerializer.cs

```diff
@@ -175,6 +164,150 @@ public override void Serialize(IIonWriter writer, object item)
+
+            return true;
+        }
+        
+        // Deserialize the given property and return bool to indicate whether the deserialized result should be used.
+        private bool TryDeserializeProperty(
+            PropertyInfo property, IIonReader reader, IonType ionType, ref object deserialized)
+        {
+            // We deserialize whether or not we ultimately use the result because values
+            // for some Ion types need to be consumed in order to advance the reader.
+            deserialized = ionSerializer.Deserialize(reader, property.PropertyType, ionType);
+            
+            if (IsReadOnlyProperty(property))
+            {
+                // property.SetValue() does not work with a readonly property.
+                // logic for handling deserializing readonly properties happens during field processing
+                // when we detect backing fields for the property.
+                return false;
+            }
+
+            return !options.IgnoreDefaults || deserialized != default;
+        }
+        
+        // Deserialize the given field and return bool to indicate whether the deserialized result should be used.
+        private bool TryDeserializeField(FieldInfo field, IIonReader reader, IonType ionType, ref object deserialized)
+        {
+            // We deserialize whether or not we ultimately use the result because values
+            // for some Ion types need to be consumed in order to advance the reader.
+            deserialized = ionSerializer.Deserialize(reader, field.FieldType, ionType);
+            
+            if (options.IgnoreReadOnlyFields && field.IsInitOnly)
+            {
+                return false;
+            }
+
+            return !options.IgnoreDefaults || deserialized != default;
+        }
+
+        private object DeserializeWithIonConstructor(ConstructorInfo ionConstructor, IIonReader reader)
+        {
+            var parameters = ionConstructor.GetParameters();
+
+            // Compute mapping between parameter names and index in parameter array so we can figure out the
+            // correct order of the constructor arguments.
+            var paramIndexMap = new Dictionary<string, int>();
+            for (int i = 0; i < parameters.Length; i++)
+            {
+                var ionPropertyName = (IonPropertyName) parameters[i].GetCustomAttribute(typeof(IonPropertyName));
+                if (ionPropertyName == null)
+                {
+                    throw new NotSupportedException(
+                        $"Parameter '{parameters[i].Name}' is not specified with the [IonPropertyName] attribute " +
+                        $"for {targetType.Name}'s IonConstructor. All constructor arguments must be annotated " +
+                        "so we know which parameters to set at construction time.");
+                }
+
+                paramIndexMap.Add(ionPropertyName.Name, i);
+            }
+
+            reader.StepIn();
```

Comment:
The below looks like it is more or less duplicated from the root `Deserialize` method.

Reply:
Duplication removed.